packages/jaeger-ui/src/model/ddg/transform-ddg-data.tsx

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  TDdgPayload,
+  TDdgService,
+  TDdgOperation,
+  TDdgPath,
+  TDdgServiceMap,
+  TDdgPathElemsByDistance,
+  TDdgTransformedDdgData,
+  PathElem,
+} from './types';
+
+export default function transformDdgData(
+  payload: TDdgPayload,
+  { service: focalService, operation: focalOperation }: { service: string; operation?: string }
+): TDdgTransformedDdgData {
+  const serviceMap: TDdgServiceMap = new Map();
+  const pathElemsByDistance: TDdgPathElemsByDistance = new Map();
+
+  const paths = payload.map(payloadPath => {
+    const path = {} as TDdgPath;
+
+    const members = payloadPath.map(({ operation: operationName, service: serviceName }, i) => {
+      if (!serviceMap.has(serviceName)) {
+        serviceMap.set(serviceName, {
+          name: serviceName,
+          operations: new Map(),
+        });
+      }
+      const service = serviceMap.get(serviceName) as TDdgService;
+
+      if (!service.operations.has(operationName)) {
+        service.operations.set(operationName, {
+          name: operationName,
+          service,
+          pathElems: [],
+        });
+      }
+      const operation = service.operations.get(operationName) as TDdgOperation;
+
+      if (
+        path.focalIdx == null &&
+        serviceName === focalService &&
+        (focalOperation == null || operationName === focalOperation)
+      ) {
+        path.focalIdx = i;
+      }
+
+      return new PathElem({ path, operation, pathIdx: i });
+    });
+
+    if (path.focalIdx == null) {
+      throw new Error('A payload path lacked the focalNode');
+    }
+
+    path.members = members;
+    members.forEach(member => {
+      if (pathElemsByDistance.has(member.distance)) {
+        (pathElemsByDistance.get(member.distance) as PathElem[]).push(member);
+      } else {
+        pathElemsByDistance.set(member.distance, [member]);
+      }
```

Comment:
Same as above, the use of a local var can help typescript along:

```tsx
let members = pathElemsByDistance.get(member.distance);
if (members) {
  members.push(member);
} else {
  pathElemsByDistance.set(member.distance, [member]);
}
```

I think typescript just doesn't have a full appreciation for what `Map#has()` offers...

Reply:
I think it's an issue with our TS config. This TS playground has no issue with the `if (!map.has(key)) map.set(key, value); map.get(key);` pattern:

https://www.typescriptlang.org/play/#src=const%20functionIndexer%3A%20Map%3C(any)%20%3D%3E%20any%2C%20number%20%3E%20%3D%20new%20Map()%3B%0D%0A%0D%0Afunction%20getFnIdx(fn%3A%20(any)%20%3D%3E%20any)%3A%20number%20%7B%0D%0A%20%20%20%20if%20(!functionIndexer.has(fn))%20%7B%0D%0A%20%20%20%20%20%20%20%20functionIndexer.set(fn%2C%20functionIndexer.size%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20functionIndexer.get(fn)%3B%0D%0A%7D

(Not sure why Github won't let me link to that using `[label](link)` even when using `ctrl + k`...)

Reply:
Though I updated to the pattern you suggested and it works well. But then shortly below this in the `while` loop it is more convoluted than this change.

Reply:
I agree that we should avoid `as` as much as possible as it generally indicates something is mistyped. Fortunately it isn't a [total sidestep](https://www.typescriptlang.org/play/#src=type%20TFoo%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Atype%20TBar%20%3D%20%7B%0D%0A%20%20%20%20bar%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20barToFoo(bar%3A%20TBar)%3A%20TFoo%20%7B%0D%0A%20%20%20%20return%20bar%20as%20TFoo%3B%0D%0A%7D) but still should be avoided.

If we could somehow "teach" TS `Map#has` and any future related shortcomings of TS that'd be ideal, but failing that the code blocks in this thread serve as good examples for helping TS.

I updated the related code, but with a `do { /* ... */ } while (/* ... */);` loop instead and it works well.